File: web/src/types/conversation/message.ts
/**
 * 消息类型定义
 *
 * 本文件定义前端使用的消息数据结构
 */
import { SessionId, ToolInput } from "../system/sdk";

export type MessageRole = "user" | "assistant" | "system" | "agent";

export interface TextContent {
  type: "text";
  text: string;
}

export interface ToolUseErrorContent {
  type: "tool_use_error";
  content: string;
}

export interface ToolUseContent {
  type: "tool_use";
  id: string;
  name: string;
  input: ToolInput;
}

export interface ToolResultContent {
  type: "tool_result";
  tool_use_id: string;
  content: string | any[];
  is_error?: boolean;
  error_code?: string | null;
}

export interface ThinkingContent {
  type: "thinking";
  thinking: string;
  signature?: string | null;
}

export interface TaskProgressContent {
  type: "task_progress";
  task_id: string;
  description: string;
  tool_use_id?: string | null;
  last_tool_name?: string | null;
  usage?: Record<string, any>;
}

export type SystemEventTone = "neutral" | "warning";
export type SystemEventIcon = "retry" | "progress" | "status";

export interface SystemEventContent {
  type: "system_event";
  content: string;
  label: string;
  tone: SystemEventTone;
  icon: SystemEventIcon;
  source_message_id: string;
  timestamp: number;
  subtype?: string;
  tool_use_id?: string | null;
}

export type ContentBlock =
  | TextContent
  | ToolUseErrorContent
  | ToolUseContent
  | ToolResultContent
  | ThinkingContent
  | TaskProgressContent
  | SystemEventContent;

export interface BaseMessage {
  message_id: string;
  session_key: string;
  room_id?: string | null;
  conversation_id?: string | null;
  agent_id: string;
  round_id: string;
  session_id?: SessionId;
  parent_id?: string;
  role: MessageRole;
  timestamp: number;
  /** 运行态投影字段；用于区分 durable / ephemeral，不来自历史回放。 */
  delivery_mode?: "durable" | "ephemeral";
}

export interface UserMessage extends BaseMessage {
  role: "user";
  content: string;
  delivery_policy?: "queue" | "guide" | "interrupt" | "auto";
}

export interface AgentMessage {
  message_id: string;
  session_key: string;
  room_id: string;
  conversation_id: string;
  sender_agent_id: string;
  content: string;
  timestamp: number;
}

export interface Usage {
  input_tokens: number;
  output_tokens: number;
  cache_creation_input_tokens?: number;
  cache_read_input_tokens?: number;
  [key: string]: any;
}

export interface ResultSummary {
  message_id?: string;
  timestamp?: number;
  subtype: "success" | "error" | "interrupted";
  duration_ms: number;
  duration_api_ms: number;
  num_turns: number;
  total_cost_usd?: number;
  usage?: Usage;
  result?: string;
  is_error: boolean;
}

/** Status for assistant messages in Room multi-agent scenarios. */
export type AssistantMessageStatus =
  | "pending"
  | "streaming"
  | "done"
  | "cancelled"
  | "error";

export interface AssistantMessage extends BaseMessage {
  role: "assistant";
  content: ContentBlock[];
  is_complete?: boolean;
  stop_reason?: "end_turn" | "max_tokens" | "stop_sequence" | "tool_use";
  model?: string;
  usage?: Usage;
  result_summary?: ResultSummary;
  /** UI-only；不是后端协议字段。 */
  stream_status?: AssistantMessageStatus;
}

export interface SystemMessageMetadata extends Record<string, any> {
  subtype?: string;
  attempt?: number;
  max_retries?: number;
  retry_delay_ms?: number;
  error_status?: string | null;
  error?: string | null;
}

export interface SystemMessage extends BaseMessage {
  role: "system";
  content: string;
  metadata?: SystemMessageMetadata;
}

export type Message = UserMessage | AssistantMessage | SystemMessage;

export type RoundLifecycleStatus =
  | "running"
  | "finished"
  | "interrupted"
  | "error";

export interface RoundStatusEventPayload {
  round_id: string;
  status: RoundLifecycleStatus;
  is_terminal: boolean;
  result_subtype?: ResultSummary["subtype"] | null;
}

export interface SessionStatusEventPayload {
  is_generating: boolean;
  running_round_ids?: string[];
  controller_client_id?: string | null;
  observer_count?: number;
  bound_client_count?: number;
}

export type StreamMessageType =
  | "message_start"
  | "content_block_start"
  | "content_block_delta"
  | "message_delta"
  | "message_stop";

export interface StreamMessage {
  message_id: string;
  session_key: string;
  room_id?: string | null;
  conversation_id?: string | null;
  agent_id: string;
  round_id: string;
  session_id?: SessionId;
  type: StreamMessageType;
  index?: number;
  content_block?: ContentBlock;
  message?: {
    model?: string;
    stop_reason?: AssistantMessage["stop_reason"];
  };
  usage?: Usage;
  timestamp: number;
}

export interface EventMessage {
  envelope_id?: string;
  protocol_version?: number;
  delivery_mode?: "durable" | "ephemeral";
  session_seq?: number;
  room_seq?: number;
  event_type:
    | "message"
    | "stream"
    | "permission_request"
    | "agent_runtime_event"
    | "workspace_event"
    | "pong"
    | "error"
    | "room_collaboration"
    | "room_member_added"
    | "room_member_removed"
    | "room_deleted"
    | "room_resync_required"
    | "session_resync_required"
    | "chat_ack"
    | "input_queue"
    | "round_status"
    | "stream_start"
    | "stream_end"
    | "stream_cancelled"
    | "permission_request_resolved"
    | "session_status";
  session_key?: string | null;
  room_id?: string | null;
  conversation_id?: string | null;
  agent_id?: string | null;
  message_id?: string | null;
  session_id?: SessionId | null;
  caused_by?: string | null;
  data: any;
  timestamp: number;
}

/** Pending agent slot from chat_ack */
export interface PendingAgentSlot {
  agent_id: string;
  msg_id: string;
  round_id?: string;
  status?: AssistantMessageStatus;
  timestamp?: number;
}

/** Room 前端占位槽位状态。 */
export interface RoomPendingAgentSlotState extends PendingAgentSlot {
  round_id: string;
  status: AssistantMessageStatus;
  timestamp: number;
}

/** chat_ack event data */
export interface ChatAckData {
  req_id: string;
  round_id: string;
  pending: PendingAgentSlot[];
  ack_timeout_ms?: number;
}

export type RoomCollaborationEventType = "agent_message" | "room_broadcast";

export interface RoomCollaborationEvent {
  event_type: "room_collaboration";
  data: {
    room_id: string;
    conversation_id: string;
    message_type: RoomCollaborationEventType;
    sender_agent_id?: string;
    content?: string;
  };
  timestamp: number;
}

export interface SystemMessageDisplayMeta {
  label: string;
  tone: "neutral" | "warning";
  icon: "retry" | "progress" | "status";
}

export function get_system_message_display_meta(
  message: SystemMessage,
): SystemMessageDisplayMeta {
  const subtype = message.metadata?.subtype;
  if (subtype === "api_retry") {
    return {
      label: "自动重试",
      tone: "warning",
      icon: "retry",
    };
  }

  if (subtype === "task_started" || subtype === "task_progress") {
    return {
      label: "执行状态",
      tone: "neutral",
      icon: "progress",
    };
  }

  if (subtype === "task_notification" || subtype === "status") {
    return {
      label: "状态更新",
      tone: "neutral",
      icon: "status",
    };
  }

  return {
    label: "系统事件",
    tone: "neutral",
    icon: "status",
  };
}

File: internal/protocol/message.go
package protocol

import (
	"encoding/json"
	"strings"
	"time"
)

// EventType 表示统一事件类型。
type EventType string

// ChatAckTimeoutMS 是客户端等待 chat_ack 的上限（毫秒）。
// 服务端不强制该窗口，但承诺在此之前回 ack；
// 前端据此设置本地超时，两侧同源避免漂移。
const ChatAckTimeoutMS = 10000

const (
	EventTypeMessage                   EventType = "message"
	EventTypeStream                    EventType = "stream"
	EventTypeChatAck                   EventType = "chat_ack"
	EventTypeRoundStatus               EventType = "round_status"
	EventTypeSessionStatus             EventType = "session_status"
	EventTypePermissionRequest         EventType = "permission_request"
	EventTypePermissionRequestResolved EventType = "permission_request_resolved"
	EventTypeAgentRuntimeEvent         EventType = "agent_runtime_event"
	EventTypeWorkspaceEvent            EventType = "workspace_event"
	EventTypeRoomMemberAdded           EventType = "room_member_added"
	EventTypeRoomMemberRemoved         EventType = "room_member_removed"
	EventTypeRoomDeleted               EventType = "room_deleted"
	EventTypeRoomResyncRequired        EventType = "room_resync_required"
	EventTypeSessionResyncRequired     EventType = "session_resync_required"
	EventTypeStreamStart               EventType = "stream_start"
	EventTypeStreamEnd                 EventType = "stream_end"
	EventTypeStreamCancelled           EventType = "stream_cancelled"
	EventTypeError                     EventType = "error"
	EventTypePong                      EventType = "pong"
)

// EventMessage 对齐前后端统一 envelope。
type EventMessage struct {
	EnvelopeID      string         `json:"envelope_id,omitempty"`
	ProtocolVersion int            `json:"protocol_version"`
	DeliveryMode    string         `json:"delivery_mode,omitempty"`
	EventType       EventType      `json:"event_type"`
	SessionKey      string         `json:"session_key,omitempty"`
	SessionSeq      *int64         `json:"session_seq,omitempty"`
	RoomID          string         `json:"room_id,omitempty"`
	RoomSeq         *int64         `json:"room_seq,omitempty"`
	ConversationID  string         `json:"conversation_id,omitempty"`
	AgentID         string         `json:"agent_id,omitempty"`
	MessageID       string         `json:"message_id,omitempty"`
	SessionID       string         `json:"session_id,omitempty"`
	CausedBy        string         `json:"caused_by,omitempty"`
	Data            map[string]any `json:"data"`
	Timestamp       int64          `json:"timestamp"`
}

// InboundWebSocketMessage 表示前端发送给服务端的基础消息。
type InboundWebSocketMessage struct {
	Type       string `json:"type"`
	SessionKey string `json:"session_key,omitempty"`
	ClientID   string `json:"client_id,omitempty"`
}

// RoundStatusData 表示 round 生命周期事件。
type RoundStatusData struct {
	RoundID       string `json:"round_id"`
	Status        string `json:"status"`
	IsTerminal    bool   `json:"is_terminal"`
	ResultSubtype string `json:"result_subtype,omitempty"`
}

// SessionStatusData 表示 session 生命周期事件。
type SessionStatusData struct {
	IsGenerating     bool     `json:"is_generating"`
	RunningRoundIDs  []string `json:"running_round_ids,omitempty"`
	ControllerClient string   `json:"controller_client_id,omitempty"`
	ObserverCount    int      `json:"observer_count,omitempty"`
	BoundClientCount int      `json:"bound_client_count,omitempty"`
}

// NewEvent 构造通用事件。
func NewEvent(eventType EventType, data map[string]any) EventMessage {
	return EventMessage{
		ProtocolVersion: 2,
		DeliveryMode:    "ephemeral",
		EventType:       eventType,
		Data:            data,
		Timestamp:       time.Now().UnixMilli(),
	}
}

// NewErrorEvent 构造错误事件。
func NewErrorEvent(sessionKey string, message string) EventMessage {
	event := NewEvent(EventTypeError, map[string]any{
		"message": message,
	})
	event.SessionKey = sessionKey
	return event
}

// NewPongEvent 构造 pong 事件。
func NewPongEvent(sessionKey string) EventMessage {
	event := NewEvent(EventTypePong, map[string]any{})
	event.SessionKey = sessionKey
	return event
}

// TypeScriptDefinitions 返回生成前端类型定义所需内容。
func TypeScriptDefinitions() string {
	const definitions = `/**
 * 由 cmd/protocol-tsgen 自动生成，请勿手改。
 */

export type EventType =
  | 'message'
  | 'stream'
  | 'chat_ack'
  | 'round_status'
  | 'session_status'
  | 'permission_request'
  | 'permission_request_resolved'
  | 'agent_runtime_event'
  | 'workspace_event'
  | 'room_member_added'
  | 'room_member_removed'
  | 'room_deleted'
  | 'session_resync_required'
  | 'room_resync_required'
  | 'stream_start'
  | 'stream_end'
  | 'stream_cancelled'
  | 'error'
  | 'pong';

export interface EventMessage {
  envelope_id?: string;
  protocol_version: number;
  delivery_mode?: string;
  event_type: EventType;
  session_key?: string;
  session_seq?: number;
  room_id?: string;
  room_seq?: number;
  conversation_id?: string;
  agent_id?: string;
  message_id?: string;
  session_id?: string;
  caused_by?: string;
  data: Record<string, unknown>;
  timestamp: number;
}

export interface RoundStatusData {
  round_id: string;
  status: string;
  is_terminal: boolean;
  result_subtype?: string;
}

export interface SessionStatusData {
  is_generating: boolean;
  running_round_ids?: string[];
  controller_client_id?: string;
  observer_count?: number;
  bound_client_count?: number;
}
`
	return definitions
}

// NewRoundStatusEvent 构造 round_status 事件。
func NewRoundStatusEvent(sessionKey string, roundID string, status string, resultSubtype string) EventMessage {
	data := map[string]any{
		"round_id":    roundID,
		"status":      status,
		"is_terminal": status == "finished" || status == "interrupted" || status == "error",
	}
	if strings.TrimSpace(resultSubtype) != "" {
		data["result_subtype"] = strings.TrimSpace(resultSubtype)
	}
	event := NewEvent(EventTypeRoundStatus, data)
	event.SessionKey = sessionKey
	return event
}

// NewChatAckEvent 构造 chat_ack 事件。
func NewChatAckEvent(sessionKey string, reqID string, roundID string, pending []map[string]any) EventMessage {
	event := NewEvent(EventTypeChatAck, map[string]any{
		"req_id":         reqID,
		"round_id":       roundID,
		"pending":        pending,
		"ack_timeout_ms": ChatAckTimeoutMS,
	})
	event.SessionKey = sessionKey
	return event
}

// NewPermissionRequestResolvedEvent 构造权限请求结束事件。
func NewPermissionRequestResolvedEvent(sessionKey string, requestID string, status string) EventMessage {
	event := NewEvent(EventTypePermissionRequestResolved, map[string]any{
		"request_id": requestID,
		"status":     strings.TrimSpace(status),
	})
	event.SessionKey = sessionKey
	return event
}

// MustJSON 将任意对象编码成紧凑 JSON 字符串。
func MustJSON(value any) string {
	payload, err := json.Marshal(value)
	if err != nil {
		return "{}"
	}
	return string(payload)
}
